File: src/python/_1448_Count_Good_Nodes_in_Binary_Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def goodNodes(self, root: TreeNode) -> int:
        def dfs(node, maxVal):
            good = 0

            if node.val >= maxVal:
                good += 1
                maxVal = node.val

            if node.left:
                good += dfs(node.left, maxVal)

            if node.right:
                good += dfs(node.right, maxVal)

            return good

        return dfs(root, root.val)

File: src/java/_1448_Count_Good_Nodes_in_Binary_Tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 * int val;
 * TreeNode left;
 * TreeNode right;
 * TreeNode() {}
 * TreeNode(int val) { this.val = val; }
 * TreeNode(int val, TreeNode left, TreeNode right) {
 * this.val = val;
 * this.left = left;
 * this.right = right;
 * }
 * }
 */
class Solution {
    private int dfs(TreeNode root, int maxVal) {
        int good = 0;

        if (root.val >= maxVal) {
            good++;
            maxVal = root.val;
        }

        if (root.left != null)
            good += this.dfs(root.left, maxVal);

        if (root.right != null)
            good += this.dfs(root.right, maxVal);

        return good;
    }

    public int goodNodes(TreeNode root) {
        return this.dfs(root, root.val);
    }
}
